base/type_intersect_exact.jl

```diff
@@ -0,0 +1,98 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module TypeIntersectExact
+
+struct OK end
+
+"""
+When `OK <: ok`, `R` is the exact type intersection. When `ok <: Union{}`, the
+exact type intersection wasn't found and `R` has no significance.
+"""
+struct Result{R, ok<:OK} end
+
+result(::Type{T}) where {T} = Result{T,       OK     }()
+failure()                   = Result{nothing, Union{}}()
+
+"""
+    get_result(::Type, ::Result)::Type
+
+Returns the exact type intersection if it was found, otherwise returns the
+fallback.
+"""
+function get_result end
+
+get_result(::Type{Fallback}, ::Result{<:Any, Union{}}) where {Fallback}    = Fallback
+get_result(::Type{Fallback}, ::Result{R,     OK     }) where {Fallback, R} = R::Type{R}
+
+"""
+    type_intersect_exact(types...)::Result
+
+Finds an exact type intersection or reports failure.
+"""
+function type_intersect_exact end
+
+type_intersect_exact() = result(Any)::Result
+
+function type_intersect_exact(@nospecialize A::Type)
+    Core.@_foldable_meta
+    result(A)::Result
+end
+
+function type_intersect_exact((@nospecialize A::Type), (@nospecialize B::Type))
+    Core.@_foldable_meta
+    if A <: B
+        result(A)
+    elseif B <: A
+        result(B)
+    else
+        let AB = typeintersect(A, B), BA = typeintersect(B, A)
+            if (AB <: A) && (AB <: B)
+                result(AB)
+            elseif (BA <: A) && (BA <: B)
+                result(BA)
+            else
+                failure()
+            end
+        end
+    end::Result
+end
+
+function type_intersect_exact(
+    (@nospecialize A::Type), (@nospecialize B::Type), (@nospecialize C::Type)
+)
+    Core.@_foldable_meta  # the loop below doesn't infer as terminating
+    candidates = let
+        AB = typeintersect(A, B)
+        BA = typeintersect(B, A)
+        AC = typeintersect(A, C)
+        CA = typeintersect(C, A)
+        BC = typeintersect(B, C)
+        CB = typeintersect(C, B)
+
+        AB_C = typeintersect(AB, C)
+        C_AB = typeintersect(C, AB)
+        BA_C = typeintersect(BA, C)
+        C_BA = typeintersect(C, BA)
+        AC_B = typeintersect(AC, B)
+        B_AC = typeintersect(B, AC)
+        CA_B = typeintersect(CA, B)
+        B_CA = typeintersect(B, CA)
+        BC_A = typeintersect(BC, A)
+        A_BC = typeintersect(A, BC)
+        CB_A = typeintersect(CB, A)
+        A_CB = typeintersect(A, CB)
+
+        (
+            A, B, C,
```

Comment:
I think you're saying that `typeintersect(A, B)` can be relied on to return the "smaller" type of the two when they're ordered? Maybe you're right, but that's not documented, and I didn't read the C implementation, so I don't know if we can rely on that. If we can rely on that behavior of `typeintersect`, perhaps we should document it.

Reply:
> I think you're saying that typeintersect(A, B) can be relied on to return the "smaller" type of the two when they're ordered?

yes. If it didn't, it would be doing a really bad job of returning an intersection.